src/http_server.c

```diff
@@ -0,0 +1,314 @@
+	}
+	snprintf(url,
+	         required_size,
+	         uri_pattern,
+	         cfg->gateway_host,
+	         cfg->gateway_port,
+	         realm,
+	         encoded_realm);
+
+	log_info("Authenticate at '%s'\n", url);
+
+end:
+	free(url);
+	free(encoded_realm);
+}
+
+// Convenience function to send a response with a user readable status message and the
+// request URL shown for debug purposes. The response is shown in the user's browser
+// after being redirected from the Fortinet Server.
+static void send_status_response(int socket, const char *userMessage)
+{
+	static const char replyHeader[] = "HTTP/1.1 200 OK\r\n"
+	                                  "Content-Type: text/html\r\n"
+	                                  "Content-Length: %lu\r\n"
+	                                  "Connection: close\r\n"
+	                                  "\r\n";
+
+	static const char replyBody[] = "<!DOCTYPE html>\r\n"
+	                                "<html><body>\r\n"
+	                                "%s" // User readable message
+	                                "</body></html>\r\n";
+
+	char *replyBodyBuffer = NULL;
+	char *replyHeaderBuffer = NULL;
+	size_t replyBodySize = snprintf(NULL, 0, replyBody, userMessage) + 1;
+
+	replyBodyBuffer = malloc(replyBodySize);
+	if (!replyBodyBuffer) {
+		log_error("malloc: %s\n", strerror(errno));
+		goto end;
+	}
+	snprintf(replyBodyBuffer, replyBodySize, replyBody, userMessage);
+
+	int replyHeaderSize = snprintf(NULL, 0, replyHeader, replyBodySize) + 1;
+
+	replyHeaderBuffer = malloc(replyHeaderSize);
+	if (!replyHeaderBuffer) {
+		log_error("malloc: %s\n", strerror(errno));
+		goto end;
+	}
+	snprintf(replyHeaderBuffer,
+	         replyHeaderSize,
+	         replyHeader,
+	         strlen(replyBodyBuffer));
+
+	// Using two separate writes here to make the code not more complicated assembling
+	// the buffers.
+	write(socket, replyHeaderBuffer, strlen(replyHeaderBuffer));
+	write(socket, replyBodyBuffer, strlen(replyBodyBuffer));
+
+end:
+	free(replyBodyBuffer);
+	free(replyHeaderBuffer);
+}
+
+static int process_request(int new_socket, char *id)
+{
+	log_info("Processing HTTP SAML request\n");
+
+	int flag = 1;
+
+	setsockopt(new_socket, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int));
+
+	// Read the request
+	char request[1024];
+	// -1 : Save one place for termination in case the request is about to
+	// fill the entire buffer.
+	ssize_t read_result = read(new_socket, request, sizeof(request) - 1);
+
+	// Check for '=id' in the response
+	// If the received request from the server is larger than the buffer,
+	// the result will not be null-terminated causing strlen to behave wrong.
+	if (read_result < 0) {
+		log_error("Bad request\n");
+		send_status_response(new_socket, "Invalid redirect response from Fortinet server. VPN could not be established.");
+		return -1;
+	}
+
+	// Safety Null-terminate
+	request[sizeof(request) - 1] = 0;
+	request[read_result] = 0;
+
+	static const char *request_head = "GET /?id=";
```

Comment:
Sorry, I missed this one:
```suggestion
	static const char request_head[] = "GET /?id=";
```